AWS SAM Project template in Rider using AWS SAM generator

sdubov opened the pull request:
## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)

## Description
First step in adding plugin support in Rider is to create a project template that we could use for running/publishing/debugging/etc. This commit provide a custom frontend template to Rider Templates wizard and run a SAM CLI project generator to create a template. There were two options in templates processing (standard dotnet new templates written from the one that is generated using SAM CLI vs. writing a custom UI in Rider Welcome Wizard and use SAM CLI generator). We went with option 2 (custom UI + generator) to unify the templates process creation and to not support template in plugin directly.

**NOTE**: This commit does not do a post-creation work (open created solution, etc.) since generated project structure is still under discussion.

## Motivation and Context
Users should be able to create a project right from Rider IDE instead selecting a proper template from Welcome Wizard instead of generating it manually from SAM CLI, generating a solution. This steps simplify the process.

## Related Issue(s)
None.

## Testing
Mac OS X. Mojave. 10.14.5
- Checked Plugin is running in Rider IDE
- Checked Welcome Wizard shows new SAM Hello World Project template
- Checked project creation under available runtimes + project structure
- Run `./gradlew check`, `./gradlew test`, `./gradlew integrationTest`

## Screenshots (if appropriate)
<img width="803" alt="RiderProjectTemplate" src="https://user-images.githubusercontent.com/6064345/58080691-56181e00-7bbc-11e9-82db-cce9f6d98f02.png">

## Checklist
- [x] I have read the **README** document
- [x] I have read the **CONTRIBUTING** document
- [x] Local run of `gradlew check` succeeds
- [x] My code follows the code style of this project
- [ ] I have added tests to cover my changes
- [x] All new and existing tests passed
- [x] A short description of the change has been added to the **CHANGELOG**

## License
I confirm that my contribution is made under the terms of the Apache 2.0 license.

abrooksv commented on jetbrains-rider/src/software/aws/toolkits/jetbrains/ui/wizard/RiderSamProjectProvider.kt:
```diff
@@ -0,0 +1,76 @@
+
+            val version = versions.sortedBy { it }.lastOrNull() ?: return defaultNetCoreRuntime
+
+            return Runtime.fromValue("dotnetcore${version.split('.').take(2).joinToString(".")}") ?: defaultNetCoreRuntime
```
`Runtime.fromValue` doesn't return null, but instead `Runtime.UNKNOWN_TO_SDK_VERSION`

We have the extension method `validOrNull` to null that out

abrooksv commented on jetbrains-rider/src/software/aws/toolkits/jetbrains/ui/wizard/RiderSamProjectGenerator.kt:
```diff
@@ -0,0 +1,135 @@
+        validationError.set(null)
+    }
+
+    private fun htmlText(s1: String, s2: String) =
```
Can you provide better names for the params?

sdubov replied:
Updated to `baseDir` and `relativePath` correspondingly.

abrooksv commented on jetbrains-core/tst/software/aws/toolkits/jetbrains/utils/rules/DotNetCodeInsightTestFixtureRule.kt:
```diff
@@ -0,0 +1,24 @@
+class DotNetCodeInsightTestFixtureRule : CodeInsightTestFixtureRule() {
+
+    override fun createTestFixture(): CodeInsightTestFixture {
+
```
Remove extra blank line

sdubov replied:
Done.

abrooksv commented on jetbrains-core/tst/software/aws/toolkits/jetbrains/utils/rules/DotNetCodeInsightTestFixtureRule.kt:
```diff
@@ -0,0 +1,24 @@
+import com.intellij.testFramework.fixtures.CodeInsightTestFixture
+import com.intellij.testFramework.fixtures.IdeaTestFixtureFactory
+
+class DotNetCodeInsightTestFixtureRule : CodeInsightTestFixtureRule() {
```
I would move this into the rider module

abrooksv commented on jetbrains-core/src/software/aws/toolkits/jetbrains/services/lambda/dotnet/DotNetRuntimeGroup.kt:
```diff
@@ -0,0 +1,22 @@
+import software.amazon.awssdk.services.lambda.model.Runtime
+import software.aws.toolkits.jetbrains.services.lambda.SdkBasedRuntimeGroupInformation
+
+class DotNetRuntimeGroup : SdkBasedRuntimeGroupInformation() {
```
I would keep this in the rider module

sdubov replied:
Done.

abrooksv commented on jetbrains-core/src/software/aws/toolkits/jetbrains/ui/wizard/SamInitSelectionPanel.java:
```diff
@@ -44,6 +44,8 @@
                                .sorted()
                                .forEach(y -> runtime.addItem(y));
 
+        runtime.setSelectedItem(this.projectSettings.runtime);
```
I am not sure looking at the existing code if this is actually safe since I only see it set the first time in the runtimeUpdate() call. It seems like it show be throwing an NPE due to the property is lateinit.

abrooksv commented on jetbrains-rider/src/software/aws/toolkits/jetbrains/ui/wizard/RiderSamProjectGenerator.kt:
```diff
@@ -0,0 +1,135 @@
+
+    private val samPanel: SamInitSelectionPanel
+
+    private val tabb: JTabbedPane
```
Call it projectStructure or something along those lines?

sdubov replied:
Renamed to `projectStructurePanel`.

abrooksv commented on jetbrains-rider/src/software/aws/toolkits/jetbrains/ui/wizard/RiderSamProjectProvider.kt:
```diff
@@ -0,0 +1,76 @@
+        override fun getKeywords() = arrayOf(name)
+
+        private fun getCurrentRuntime(): Runtime {
+
```
remove blank line

sdubov commented:
@abrooksv, I've fixed your comments above. In addition, I've added a logic to generate solution file and a proper project structure for SAM Project Template. Currently, we add two projects - `HelloWorld` and `HelloWorld.Test` into a solution, and add a separate Solution folder `SolutionItems` with the rest of generated files (`template.yaml`, `omnisharp.json`, `README.md`, `.gitignore`). 

To add a project structure I need to call Rider API added to latest 192 EAP SDK. So, I've switched the plugin to 192 EAP release and made some updates in calls (fixed deprecated calls, drop ones that are not used, etc.). 

Note: The `./gradlew check` linter found a deprecated call in `RemoteLambdaState.kt` class:  `SearchScopeProvider.createSearchScope`. If update it to a suggested API `GlobalSearchScopes.executionScope`, then it produce another error about incompatible versions (since the API is added since `191.7141.2` that is conflict with plugin `since` version that is 183...). I've just suppress this warning for now.

Image with latest updates:
<img width="802" alt="RiderProjectTemplateSamApp" src="https://user-images.githubusercontent.com/6064345/58438543-095ea680-80d8-11e9-9a85-da160383d755.png">

abrooksv commented:
Since this requires 192 for Rider, and since the `since` attribute in plugin.xml is global for all IDEs, this forces us to re-baseline on 192, correct?

sdubov commented:
That's correct. To support Rider will need some more new APIs available in SDK (e.g. I've just recently added an ability to validate handler for lambda run configuration that require ReSharper PSI lookup to correctly define if method exists since we cannot use Idea PSI in this case).

abrooksv commented:
I opened https://github.com/aws/aws-toolkit-jetbrains/pull/1000 against master to enable 192 so that we can go ahead and get EAP users able to use the toolkit.

abrooksv commented:
@sdubov I have no issues with this PR as it stands now, do you want me to merge it into the branch or are there more changes for project template support?

sdubov commented:
Please, go ahead. I have no more changes for project templates for now. I will send another PR with Lambda Local Run right after.

abrooksv commented:
Okay, merged into https://github.com/aws/aws-toolkit-jetbrains/commits/feature/rider-support

Please PR against the feature branch since our CI jobs filter on that

sdubov commented:
Got it. Thank you.
